signer/mar/mar.go

```diff
@@ -0,0 +1,218 @@
+package mar
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	"go.mozilla.org/autograph/signer"
+	margo "go.mozilla.org/mar"
+
+	"github.com/ThalesIgnite/crypto11"
+)
+
+const (
+	// Type of this signer is "mar"
+	Type = "mar"
+)
+
+// MARSigner holds the configuration of the signer
+type MARSigner struct {
+	signer.Configuration
+	signingKey    crypto.PrivateKey
+	publicKey     crypto.PublicKey
+	rand          io.Reader
+	defaultSigAlg uint32
+}
+
+// New initializes a mar signer using a configuration
+func New(conf signer.Configuration) (s *MARSigner, err error) {
+	s = new(MARSigner)
+
+	if conf.Type != Type {
+		return nil, errors.Errorf("mar: invalid type %q, must be %q", conf.Type, Type)
+	}
+	s.Type = conf.Type
+
+	if conf.ID == "" {
+		return nil, errors.New("mar: missing signer ID in signer configuration")
+	}
+	s.ID = conf.ID
+
+	if conf.PrivateKey == "" {
+		return nil, errors.New("mar: missing private key in signer configuration")
+	}
+	s.PrivateKey = conf.PrivateKey
+	s.signingKey, err = conf.GetPrivateKey()
+	if err != nil {
+		return nil, errors.Wrap(err, "mar: failed to parse private key")
+	}
+
+	// do a bunch of type assertions to get the public key and default signature algorithm
+	// up front so we don't have to clutter the rest of the code with this
+	switch s.signingKey.(type) {
+	case *rsa.PrivateKey:
+		s.publicKey = s.signingKey.(*rsa.PrivateKey).Public()
+		publicKeyBytes, err := x509.MarshalPKIXPublicKey(&s.signingKey.(*rsa.PrivateKey).PublicKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "mar: failed to asn1 marshal rsa public key")
+		}
+		s.PublicKey = base64.StdEncoding.EncodeToString(publicKeyBytes)
+		s.defaultSigAlg = margo.SigAlgRsaPkcs1Sha384
+		s.rand = rand.Reader
+
+	case *crypto11.PKCS11PrivateKeyRSA:
+		s.publicKey = s.signingKey.(*crypto11.PKCS11PrivateKeyRSA).Public()
+		publicKeyBytes, err := x509.MarshalPKIXPublicKey(s.signingKey.(*crypto11.PKCS11PrivateKeyRSA).PubKey.(*rsa.PublicKey))
+		if err != nil {
+			return nil, errors.Wrap(err, "mar: failed to asn1 marshal crypto11 rsa public key")
+		}
+		s.PublicKey = base64.StdEncoding.EncodeToString(publicKeyBytes)
+		s.defaultSigAlg = margo.SigAlgRsaPkcs1Sha384
+		s.rand = new(crypto11.PKCS11RandReader)
+
+	case *ecdsa.PrivateKey:
+		s.publicKey = s.signingKey.(*ecdsa.PrivateKey).Public()
+		publicKeyBytes, err := x509.MarshalPKIXPublicKey(&s.signingKey.(*ecdsa.PrivateKey).PublicKey)
+		if err != nil {
+			return nil, errors.Wrap(err, "mar: failed to asn1 marshal ecdsa public key")
+		}
+		s.PublicKey = base64.StdEncoding.EncodeToString(publicKeyBytes)
+		switch s.publicKey.(*ecdsa.PublicKey).Params().Name {
+		case elliptic.P256().Params().Name:
+			s.defaultSigAlg = margo.SigAlgEcdsaP256Sha256
+		case elliptic.P384().Params().Name:
+			s.defaultSigAlg = margo.SigAlgEcdsaP384Sha384
+		}
```

Comment:
Does this error when an unrecognized name is provided?

Reply:
Yep, and I added a test in 06a875df